docs/DesignDocs/SGX2-EDMM-driver-interface.md

```diff
@@ -0,0 +1,137 @@
+SGX EDMM Driver Interface Design
+=====================================
+
+## Motivation
+
+This document describes possible driver interfaces to facilitate discussions among SGX runtime implementors (e.g., https://github.com/openenclave/openenclave/pull/3639) on supporting different SGX EDMM flows.
+
+Although interfaces described here are inspired to be as likely as possible a candidate for future kernel adoption, they are not intended to be a proposal for kernel implementation and are assumed to be implemented as an OOT driver. We hope from discussions enabled by this document, requirements and usage models can be identified to help shape future kernel interfaces. 
+
+## Basic EDMM flows
+
+SGX EDMM instructions support dynamic EPC page allocation/deallocation for enclaves and page property modification post-EINIT.  Following are the basic EDMM flows on which other more advanced usages of EDMM can be built.
+
+**Note:** The term "kernel" and "kernel space" are used in this document when general kernel space actions are described whether implemented in an OOT driver or in kernel tree. Kernel specific implementation details will be explicitly stated as "future kernel" or "kernel patches".  And implementation details such as OCalls issued by enclaves, ETRACK and IPI issued in kernel are generally omitted for brevity.
+
+- Allocate a new page at an address in ELRANGE of an enclave. 
+  - This can be an explicit syscall or triggered by #PF  when an unavailable page is accessed.
```

Comment:
From the point of view of emulating a Linux environment, it would be helpful to have an option to only allocate pages on explicit syscalls and not implicitly when triggered by a page fault. A Linux application may have a SIGSEGV handler, and may handle a page fault in a way that doesn't result in a page mapping being added at that address.

Reply:
Can you specify the exact targeted call flow?
Note for large block of EPC allocation, mmap(..., MAP_POPULATE) proposed below can be used to instruct kernel to allocate EPC as immediately as possible. However, I think kernel should make the final decision on when exactly the physical EPC pages are committed.